gpiozero/mixins.py

```diff
@@ -266,17 +300,141 @@ def _fire_events(self):
                 self._active_event.set()
             else:
                 self._inactive_event.set()
-        else:
-            if not old_state and new_state:
+        elif old_state != new_state:
+            self._last_changed = time()
+            if new_state:
                 self._inactive_event.clear()
                 self._active_event.set()
-                if self.when_activated:
-                    self.when_activated()
-            elif old_state and not new_state:
+                self._fire_activated()
+            else:
                 self._active_event.clear()
                 self._inactive_event.set()
-                if self.when_deactivated:
-                    self.when_deactivated()
+                self._fire_deactivated()
+
+
+class HoldMixin(EventsMixin):
+    def __init__(self, *args, **kwargs):
+        super(HoldMixin, self).__init__(*args, **kwargs)
+        self._when_held = None
+        self._held_from = None
+        self._hold_time = 1
+        self._hold_repeat = False
+        self._hold_thread = HoldThread(self)
+
+    def close(self):
+        if self._hold_thread:
+            self._hold_thread.stop()
+            self._hold_thread = None
+        try:
+            super(HoldMixin, self).close()
+        except AttributeError:
+            pass
+
+    def _fire_activated(self):
+        super(HoldMixin, self)._fire_activated()
+        self._hold_thread.holding.set()
+
+    def _fire_deactivated(self):
+        self._held_from = None
+        super(HoldMixin, self)._fire_deactivated()
+
+    def _fire_held(self):
+        if self.when_held:
+            self.when_held()
+
+    @property
+    def when_held(self):
+        """
+        The function to run when the device has remained active for
+        :attr:`hold_time` seconds.
+
+        This can be set to a function which accepts no (mandatory) parameters,
+        or a Python function which accepts a single mandatory parameter (with
+        as many optional parameters as you like). If the function accepts a
+        single mandatory parameter, the device that activated will be passed
+        as that parameter.
+
+        Set this property to ``None`` (the default) to disable the event.
+        """
+        return self._when_held
+
+    @when_held.setter
+    def when_held(self, value):
+        self._when_held = self._wrap_callback(value)
+
+    @property
+    def hold_time(self):
+        """
+        The length of time (in seconds) to wait after the device is activated,
+        until executing the :attr:`when_held` handler. If :attr:`hold_repeat`
+        is True, this is also the length of time between invocations of
+        :attr:`when_held`.
+        """
+        return self._hold_time
+
+    @hold_time.setter
+    def hold_time(self, value):
+        self._hold_time = value
```

Comment:
Hmm, should also check for a negative value (looking at the setter for source_delay which is similar). I'm going to be slightly naughty and rejig some of the exceptions as well because now a lot of functions are devolved into mixins they're really not specific to GPIODevice anymore so they shouldn't be GPIODeviceError descendants but GPIOZeroError descendants (yeah, okay, it's a compatibility break but I'll be amazed if anyone notices it :)